core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1967,39 +1967,138 @@ private[spark] object Utils extends Logging {
       conf: SparkConf,
       serviceName: String = ""): (T, Int) = {
```

Comment:
If i didnt do it this way then this method would have been similar but not the same with the pre-existing method in that file. Then it would have been be a problem of redundancy. The logic is fairly complex and it is the minimum i could use to cover all cases.